flips/20210916-nft-metadata.md

```diff
@@ -0,0 +1,354 @@
+you want to support. This will often lead to duplicated contracts and code and just lots 
+of things to maintain. This standard will enable you to create a single Auction contract 
+and define the Type of views you need to expose in your NFT in order to support the contract. 
+So it will enable a design-by-contract pattern where the creator of a contract can define 
+the views it requires and an NFT can tell what views they export. 
+
+All NFTs and contracts using NFTs on Flow are affected by this problem. This is most 
+visible in the number of contracts that Viv3 needs in order to support all their different 
+sellable items. With all the oncoming marketplaces/auction houses and things like that 
+coming to Flow a good standard for NFTs will help everybody. 
+
+## User Benefit
+We feel that a standard that authors can implement that enables easy consumption of metadata 
+will lead to more applications that integrate NFTs. This will make NFTs intrinsically more 
+useful and valuable to the end users holding the NFTs.
+
+A user will be much more likely to transfer their NFT to other marketplaces and sell their NFT 
+in different ways with this solution. Without it all marketplaces need to hardcode support 
+for all NFTs directly. 
+
+It will also open up the possibility to standardize on some shared structs like Royalty 
+and SaleItem/Auction.
+
+## Design Proposal
+
+The core of this proposal is to add the following interface and have the NFT resource implement it
+```
+pub resource interface ViewResolver {
+  pub fun getViews() : [Type]
+  pub fun resolveView(_ view:Type): AnyStruct
+}
+```
+
+We will also add default implementations for these methods so that no compatibility will be broken 
+for existing NFT solutions. The new NFT interface could look something like this: 
+```
+pub resource NFT: INFT, ViewResolver {
+    pub let id: UInt64
+    pub fun getViews(): [Type] {
+       return []
+    }
+    pub fun resolveView(_ type: Type): AnyStruct {
+        return nil
+    }
+}
+```
+
+We propose that all metadata be expressed in Cadence types. The `pub fun getViews(): [Type]` 
+method allows for discovery of the various types of metadata supported by the NFT by returning a `Type` 
+for each supported type of metadata. An instance of any supported type could then be retrieved by 
+calling the `pub fun resolveView(_ type: Type): AnyStruct` method passing the desired type in as 
+an argument.
+
+This proposal by no means precludes the use of member variable on the NFT resource itself. In fact,
+ideally all immutable data for an NFT is stored as immutable members of the NFT resource and views 
+are resolved and created by accessing those immutable members.
+
+### Supporting other resources besides NFTs
+In order to support any other resource and resource collections, like marketplaces and fungible tokens, the following new interface is proposed:
+```
+pub resource interface ViewResolverCollection {
```

Comment:
should this implement `ViewResolver` too?  It's not very useful as is, the IDs are basically irrelevant to anyone using this - they would have to iterate through every resolve in the collection to find one that supports the `Type` that they are trying to get a view for. If this instead implemented `ViewResolver` then all of that could happen internally and potentially have a more optimized implementation (more optimized than being forced to iterate until you find one that supports your desired view).

Reply:
My thinking has been here that all NFTs in the same collection would share the same views. But i can see where that might not be the case. 

Maybe add a method that allow you to fetch all ids that have a given view?

Reply:
As @briandilley said `ViewResolver` can be better, so it can return `[Display]` for example as supported type, and when queried, can return resolved for each item. ( though in that case gas limit can be a problem )

Reply:
Because `ViewResolver` is an interface, it could basically be implemented in any way the programmer wants - I'm not seeing the benefit of the `ViewResolverCollection` (at least as part of the standard anyway).

Reply:
Without ViewResolverCollection we have no way of reasoning over collections of resources that are not NFTs where the resources share the same views. As an example NFTStorefront.

Reply:
I think gasLimit on large collections is a big concern too, you cannot query 1000 items for a Type.